feat: Add goal lines to trends + smarter default alerts

rafaeelaudibert opened the pull request:
## Problem

When creating an insight from Web Vitals we are copying the thresholds over but they are not editable. Similarly, we'd love if people could easily set up alerts for web vitals with sane values.

## Changes

- Goal lines can be added to Trends insights!
- Alerts are smarter on creation, they'll inherit the smallest goal line value

https://www.loom.com/share/442e793b77694b8bb66e871bfb084c24?sid=5e7a71bc-5161-473a-b36f-1f4ba113e76f

## Does this work well for both Cloud and self-hosted?

Yes, sir

## How did you test this code?

Manually

rafaeelaudibert committed:
ui: Fix web vitals card color

Let's make it two colors to improve visibility

rafaeelaudibert committed:
feat: Allow goal lines to be displayed/hidden on data visualization

rafaeelaudibert committed:
feat: Add goal lines to `Trends` queries

These already existed and were being used by Web Vitals, but they're now available to be edited in the UI

rafaeelaudibert committed:
feat: Use goal lines as initial value for alerts

It's common that people will create goal lines as an indication of whether they're above the value they want to be. Because of that, when setting up a new alert from the insights page we'll now use the smallest goal line as the default value for the alert

rafaeelaudibert commented:
@mariusandra Pinging you because you're the kea resident. If you look at my video you'll see we're querying every time I change a value in the goal line input (including the label name). Any suggestion on how we could delay updating the `querySource` to avoid the query going out so often?

rafaeelaudibert commented on frontend/src/queries/nodes/DataVisualization/Components/DisplayTab.tsx:
```diff
@@ -91,34 +92,48 @@ export const DisplayTab = (): JSX.Element => {
-                            value={(goalLine.value ?? 0).toString()}
+                            value={value.toString()}
                             inputMode="numeric"
                             onChange={(value) => updateGoalLine(goalLineIndex, 'value', parseInt(value))}
```
This is handled inside `updateGoalLine`

rafaeelaudibert commented on frontend/src/queries/nodes/DataVisualization/Components/DisplayTab.tsx:
```diff
@@ -91,34 +92,48 @@ export const DisplayTab = (): JSX.Element => {
+                <LemonLabel className="mb-1">Goals</LemonLabel>
+
+                {goalLines.map(({ label, value = 0, displayLabel = true }, goalLineIndex) => (
                     <div className="flex flex-1 gap-1 mb-1" key={`${goalLineIndex}`}>
```
Items aren't reordered, and we make sure to keep indexes consistent

rafaeelaudibert commented on frontend/src/lib/components/Alerts/alertFormLogic.ts:
```diff
@@ -40,26 +47,41 @@ export interface AlertFormLogicProps {
+    insightVizDataLogicProps?: InsightLogicProps
+}
+
+const getThresholdBounds = (goalLines?: GoalLine[] | null): InsightsThresholdBounds => {
+    if (goalLines == null || goalLines.length == 0) {
+        return {}
+    }
+
+    // Simple assumption that the alert should be triggered when the first/smallest goal line is crossed
+    const smallerValue = Math.min(...goalLines.map((line) => line.value))
+    return { upper: smallerValue }
 }
```
I've literally documented the assumption :)

rafaeelaudibert commented on frontend/src/lib/components/Alerts/alertFormLogic.ts:
```diff
@@ -40,26 +47,41 @@ export interface AlertFormLogicProps {
                 ({
                     id: undefined,
-                    name: '',
+                    name: values.goalLines && values.goalLines.length > 0 ? `Crossed ${values.goalLines[0].label}` : '',
```
``label` is a required field...

rafaeelaudibert commented on frontend/src/queries/nodes/InsightViz/EditorFilters.tsx:
```diff
@@ -271,6 +272,16 @@ export function EditorFilters({ query, showing, embedded }: EditorFiltersProps):
                     : null,
             ]),
         },
+        {
+            title: 'Goals',
+            editorFilters: [
+                {
+                    key: 'goal-lines',
+                    position: 'right',
+                    component: GoalLines,
+                },
+            ],
+        },
```
Excellent catch

rafaeelaudibert committed:
fix: Display goal lines on trends only

rafaeelaudibert committed:
fix

rafaeelaudibert committed:
types: Fix unreasonable TS error

mariusandra commented:
> @mariusandra Pinging you because you're the kea resident. If you look at my video you'll see we're querying every time I change a value in the goal line input (including the label name). Any suggestion on how we could delay updating the `querySource` to avoid the query going out so often?

I think the quickest win you can do is to update the value `onBlur`, instead of `onChange`. However even then we should not need to make a request. Sadly I don't know what the current `insightLogic`, caching and reloading situation is. You'll have to ask someone from Product Analytics.

In the olden days, I had a dream, where a query would look like this:

```ts
const query: InsightVizNode = {
    kind: NodeKind.InsightVizNode,
    source: {
        kind: NodeKind.TrendsQuery,
        series: [
            {
                event: '$pageview',
                name: '$pageview',
                kind: NodeKind.EventsNode,
                math: BaseMathType.UniqueUsers,
            },
        ],
        breakdownFilter: {
            breakdown: `$feature/${featureFlag.key}`,
            breakdown_type: 'event',
        },
    },
}
```

and in there, the `TrendsQuery` would contain everything that needed a server reload, whereas all "client side" attributes would be added on the `InsightVizNode`. We'd probably make a custom `TrendsVizNode` for all the attributes.

Thus this code:

```ts
actions.updateQuerySource({
    trendsFilter: {
        ...querySource?.trendsFilter,
        goalLines: values.goalLines,
    },
} as Partial<TrendsQuery>)
```

would become:

```ts
actions.updateQuery({
    goalLines: values.goalLines,
    ...query
} as Partial<TrendsVizNode>)
```

However I believe such a distinction is not made today... and I'm not sure if there's anything in place to say "shouldn't reload".

rafaeelaudibert commented:
> I think the quickest win you can do is to update the value `onBlur`, instead of `onChange`. However even then we should not need to make a request. Sadly I don't know what the current `insightLogic`, caching and reloading situation is. You'll have to ask someone from Product Analytics.

Tried that, but it doesn't work because our `LemonInput`s is controllable. Having a local `useState` for those values is very contrived because we've got a list of goal lines. I believe I'll keep it like this for now, and we can come back and optimize this eventually.

> 
> In the olden days, I had a dream, where a query would look like this:
> 
> ```ts
> const query: InsightVizNode = {
>     kind: NodeKind.InsightVizNode,
>     source: {
>         kind: NodeKind.TrendsQuery,
>         series: [
>             {
>                 event: '$pageview',
>                 name: '$pageview',
>                 kind: NodeKind.EventsNode,
>                 math: BaseMathType.UniqueUsers,
>             },
>         ],
>         breakdownFilter: {
>             breakdown: `$feature/${featureFlag.key}`,
>             breakdown_type: 'event',
>         },
>     },
> }
> ```
> 
> and in there, the `TrendsQuery` would contain everything that needed a server reload, whereas all "client side" attributes would be added on the `InsightVizNode`. We'd probably make a custom `TrendsVizNode` for all the attributes.
> 
> Thus this code:
> 
> ```ts
> actions.updateQuerySource({
>     trendsFilter: {
>         ...querySource?.trendsFilter,
>         goalLines: values.goalLines,
>     },
> } as Partial<TrendsQuery>)
> ```
> 
> would become:
> 
> ```ts
> actions.updateQuery({
>     goalLines: values.goalLines,
>     ...query
> } as Partial<TrendsVizNode>)
> ```
> 
> However I believe such a distinction is not made today... and I'm not sure if there's anything in place to say "shouldn't reload".

Oh, absolutely. The catch here is that the backend needs the `trendsFilter` values to decide how it should group the values (area vs. line, for example) but we're reusing it for way too much stuff, including `goalLines` which aren't really a requirement in the backend - it's not limited to my addition, though, some other fields aren't used in the backend but are being passed in.

robbie-c commented:
I'll let @PostHog/team-product-analytics handle this, but if it gets to weds without a review I'll take another look :)

thmsobrmlr reviewed (changes requested):
Great work @rafaeelaudibert !

Changes look good to me code wise, aside from a question I put inline. I'm only a bit worried about further complicating the insights UI, as people are already overwhelmed by it now. Also it moves the chart further down and it's now completely below the fold on a laptop. What about moving the goal lines to a button next to the alerts one? 

IMO we could even move both into the "..."-menu at the top in a separate section. We wanted to have alerts more accessible in the beginning to help feature adoption though.

rafaeelaudibert commented:
@thmsobrmlr 

> You can add the goalLines key to the properties here [master/frontend/src/scenes/insights/utils/queryUtils.ts#L150](https://github.com/PostHog/posthog/blob/master/frontend/src/scenes/insights/utils/queryUtils.ts?rgh-link-date=2025-02-12T10%3A12%3A07Z#L150). This will ignore it when deciding if a new query should be made to the backend. Other than that you could add a debounce to the input, though up to you to decide if that feels nicer than the instantaneous moving of the goal line.

Ah, that's great. It's not easy to find, but it definitely does the job. I'll keep it without debouncing for now.

> I'm only a bit worried about further complicating the insights UI, as people are already overwhelmed by it now. Also it moves the chart further down and it's now completely below the fold on a laptop. What about moving the goal lines to a button next to the alerts one?

> IMO we could even move both into the "..."-menu at the top in a separate section. We wanted to have alerts more accessible in the beginning to help feature adoption though.

Ah, interesting. It appears to the left for me because of my screen size I suppose? I'll see what I can do here, might simplify the UI a little. I don't want to put it at the top because this is customizing **the query**, while Alerts is something extraneous to the query, so it makes sense to live somewhere else.

rafaeelaudibert committed:
fix: Stop rerendering on `goalLines` changes

rafaeelaudibert commented:
@thmsobrmlr https://github.com/PostHog/posthog/blob/master/frontend/src/scenes/insights/utils/queryUtils.ts#L129 mentions we need to keep the backend in sync with that code. It's linking to the wrong place, but I'll assume it's meant to link to this:
![image](https://github.com/user-attachments/assets/18f8c413-cbe2-466d-b674-3e8290b6e935)

It's not up-to-date anymore, would you like me to fix it, or is it intended?

rafaeelaudibert committed:
feat: Optimize layout for smaller laptop

In some smaller laptops the chart is under the fold which isn't ideal. I've moved a lot of the options under an "advanced options" dropdown. I've included the new goal lines but also some existing settings that don't need to be toggled often.

I've revamped the types as well because this is was extremely outdated. There's no more concept of left/right, nor anything to do with count.

rafaeelaudibert committed:
fix: Fix web vitals colors

rafaeelaudibert committed:
fix: Revert back to left/right differentiation

I had a feature flag locally that messed up my visualization. I've now removed it and I can clearly see I messed up in my most recent commit

rafaeelaudibert commented:
@thmsobrmlr I've hidden the goal lines + some existing settings behind a dropdown. You can see that in the new snapshots - or test locally, of course. Do you think this is mergeable now?

github-actions[bot] committed:
Update UI snapshots for `webkit` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

robbie-c commented on frontend/__snapshots__/scenes-app-insights--funnel-historical-trends-edit--light.png:
Is this just a flaky screenshot or is this a real change? IMO the older version is better

robbie-c commented on frontend/__snapshots__/scenes-app-insights--trends-number-edit--dark.png:
What does it mean to have a goal line on a Number?

robbie-c commented on frontend/src/queries/nodes/InsightViz/EditorFilters.tsx:
```diff
@@ -287,78 +290,58 @@ export function EditorFilters({ query, showing, embedded }: EditorFiltersProps):
-                    position: 'right',
                     component: SamplingFilter,
                 },
+                isTrends && {
```
IMO this should only appear for certain visualizations, as some (like number, pie chart) don't make sense

rafaeelaudibert replied:
That makes sense. I meant to do this and ended up forgetting, on it

robbie-c commented on frontend/src/queries/nodes/InsightViz/PropertyGroupFilters/PropertyGroupFilters.tsx:
```diff
@@ -54,11 +54,27 @@ export function PropertyGroupFilters({
-                        query={query}
-                        setQuery={setQuery as (node: InsightQueryNode) => void}
-                    />
+                    <div className="flex flex-1 gap-2 flex-row space-between">
```
Ah yeah - I commented on this screenshot.

What's the rationale for this change? In the screenshot examples it forces the label for InsightTestAccountFilter onto two lines which IMO looks weird

robbie-c replied:
I will defer to @thmsobrmlr and @PostHog/team-product-analytics on this :)

rafaeelaudibert committed:
fix: Display goal lines on line graphs only

github-actions[bot] committed:
Update UI snapshots for `webkit` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

thmsobrmlr commented on frontend/src/queries/nodes/WebVitals/WebVitalsProgressBar.tsx:
```diff
@@ -28,9 +28,9 @@ export function WebVitalsProgressBar({ value, metric }: WebVitalsProgressBarProp
                 style={{
                     width: `${goodWidth}%`,
-                    backgroundColor: band === 'good' ? WEB_VITALS_COLORS.good : undefined,
+                    backgroundColor: band === 'good' ? `${WEB_VITALS_COLORS.good} !important` : undefined,
```
I'm not quite sure why these changes are included in this PR, but I trust you to verify whether they are intentional.

rafaeelaudibert replied:
Yep, that was a fly-by

thmsobrmlr reviewed (approved):
Great work @rafaeelaudibert! Lot's of improvements in here.

There is one thing blocking this from merging: The flag for determining wether the `GoalLines` filter should be displayed doesn't work reliably. For example the filter isn't present for a new insight. This is because the flag doesn't take defaults into account. I trust you to fix this though and am handing out a  ✅  already.

> Ah, interesting. It appears to the left for me because of my screen size I suppose? I'll see what I can do here, might simplify the UI a little. I don't want to put it at the top because this is customizing the query, while Alerts is something extraneous to the query, so it makes sense to live somewhere else.

Oh, I see! You seem to have feature flag `insight-horizontal-controls` enabled, likely due to syncing all feature flags. That flag is an experiment and currently not rolled out to users - they'll see the version with the controls on top (just funnels are horizontal by default). The horizontal layout is also used for very small (mobile) screen sizes.

> @thmsobrmlr [`master`/frontend/src/scenes/insights/utils/queryUtils.ts#L129](https://github.com/PostHog/posthog/blob/master/frontend/src/scenes/insights/utils/queryUtils.ts#L129) mentions we need to keep the backend in sync with that code. It's linking to the wrong place, but I'll assume it's meant to link to this: ![image](https://private-user-images.githubusercontent.com/32079912/412541686-18f8c413-cbe2-466d-b674-3e8290b6e935.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk1NjU1NDIsIm5iZiI6MTczOTU2NTI0MiwicGF0aCI6Ii8zMjA3OTkxMi80MTI1NDE2ODYtMThmOGM0MTMtY2JlMi00NjZkLWI2NzQtM2U4MjkwYjZlOTM1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjE0VDIwMzQwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM3ZDY0NjBhYjU3ZTE2YzE5NzgxMmI4NGEwMDg1ODcwMWRkZGM0ZjlmY2MxMTk1ZDI2NDMwYjdhOWUyM2E3OTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.8GpOfMXQ6yVe2s-eDxA19dHrfWGnHRuqChDnclri6KY)
> 
> It's not up-to-date anymore, would you like me to fix it, or is it intended?

Yup, they seem to have gotten out of sync. Feel free to fix it, or also just merge this PR and I'll do that separately.

> What's the rationale for this change? In the screenshot examples it forces the label for InsightTestAccountFilter onto two lines which IMO looks weird.

I think the filters look better in the left-to-right layout (most insights), but don't look good for the funnels insight. Would be good to have the original layout there. Not a blocker for merging though.

thmsobrmlr commented:
Didn't see that automatic merging is enabled @rafaeelaudibert and don't think the issues I mentioned warrant a revert. Would be great if you can follow up with a fix though.

rafaeelaudibert commented:
@thmsobrmlr Oh yikes, my bad with the automerge. I'll follow up with the small fixes :). Thank you!
